gateway/src/endpoints/feedback.rs

```diff
@@ -269,14 +276,175 @@ async fn write_boolean(
+        match value {
+            DemonstrationContentBlock::Text(text) => Ok(ContentBlock::Text(text)),
+            DemonstrationContentBlock::ToolCall(tool_call) => {
+                Ok(ContentBlock::ToolCall(tool_call.try_into()?))
+            }
+        }
+    }
+}
+
+async fn get_function_name_from_inference_id(
+    inference_id: Uuid,
+    connection_info: &ClickHouseConnectionInfo,
+) -> Result<String, Error> {
+    let query = format!(
+        "SELECT function_name FROM InferenceById WHERE id = '{}'",
+        inference_id
+    );
+    let function_name = connection_info.run_query(query).await?.trim().to_string();
+    if function_name.is_empty() {
+        return Err(Error::InvalidRequest {
+            message: "Inference ID does not exist".to_string(),
+        });
+    }
+    Ok(function_name)
+}
+
+// Validates that the demonstration is correct.
+// For chat functions, the value should be a string or a list of valid content blocks.
+// For json functions, the value is validated against the output schema. If it passes,
+// we construct the usual {"raw": str, "parsed": parsed_value} object, serialize it, and return it.
+async fn validate_parse_demonstration(
+    function_config: &'static FunctionConfig,
+    tools: &'static HashMap<String, StaticToolConfig>,
+    value: &Value,
+) -> Result<String, Error> {
+    match function_config {
+        FunctionConfig::Chat(chat_config) => {
+            // For chat functions, the value should either be a string or a list of valid content blocks.
+            let content_blocks = match value {
+                Value::String(s) => {
+                    vec![DemonstrationContentBlock::Text(Text {
+                        text: s.to_string(),
+                    })]
+                }
+                Value::Array(content_blocks) => content_blocks
+                    .iter()
+                    .map(|block| {
+                        serde_json::from_value::<DemonstrationContentBlock>(block.clone()).map_err(
+                            |e| Error::InvalidRequest {
+                                message: format!("Invalid demonstration content block: {}", e),
+                            },
+                        )
+                    })
+                    .collect::<Result<Vec<DemonstrationContentBlock>, Error>>()?,
+                _ => {
+                    return Err(Error::InvalidRequest {
+                        message: "Demonstration must be a string".to_string(),
+                    });
+                }
+            };
+            let tool_config = ToolCallConfig::new(
+                &chat_config.tools,
+                &chat_config.tool_choice,
+                chat_config.parallel_tool_calls,
+                tools,
+                DynamicToolParams {
+                    allowed_tools: None,
+                    additional_tools: None,
+                    tool_choice: None,
+                    parallel_tool_calls: None,
+                },
```

Comment:
Maybe we should allow demonstrations with dynamic tools. It can be a new issue though (but should be recorded).